src/routes/v2/projects.rs

```diff
@@ -2558,3 +2561,214 @@ pub async fn delete_from_index(
+                value: None,
+            });
+        } else {
+            entry.content = Some(atom_syndication::Content {
+                base: Some(link.clone()),
+                content_type: Some("html".to_string()),
+                value: Some(rendered_changelog.to_string()),
+                lang: None,
+                src: None,
+            });
+        }
+        entry.links.push(atom_syndication::Link {
+            href: link.clone(),
+            rel: "alternate".to_string(),
+            mime_type: Some("text/html".to_string()),
+            title: Some(version.name.to_string()),
+            hreflang: None,
+            length: None,
+        });
+        entry.authors = match users.iter().find(|u| u.id == version.author_id) {
+            Some(u) => vec![atom_syndication::Person {
+                name: u.username.clone(),
+                uri: Some(format!("{site_url}/user/{}", &u.username)),
+                email: None,
+            }],
+            None => authors.clone(), // this should be unreachable
+        };
+        entry.categories = version
+            .loaders
+            .iter()
+            .map(|l| &l.0)
+            .chain(version.game_versions.iter().map(|gv| &gv.0))
+            .map(|category| atom_syndication::Category {
+                label: Some(category.to_string()),
+                term: category.to_string(),
+                scheme: None,
+            })
+            .collect();
+        feed.entries.push(entry);
+    }
+    feed.authors = authors;
+    let mut out = vec![];
+    match feed.write_with_config(
+        &mut out,
+        atom_syndication::WriteConfig {
+            write_document_declaration: true,
+            indent_size: Some(4),
+        },
+    ) {
+        Ok(_) => (),
+        Err(e) => return Err(ApiError::Xml(format!("failed to generate feed: {e}"))),
+    }
+    let checksum = {
+        use hex::ToHex;
+        use sha2::Digest;
+
+        let mut hasher = sha2::Sha256::new();
+        hasher.update(&out);
+        hasher.finalize().encode_hex()
+    };
+    let mut response = HttpResponse::Ok();
+    response.content_type("application/atom+xml");
+    response.append_header((header::ACCESS_CONTROL_ALLOW_ORIGIN, "*"));
+    response.append_header(header::CacheControl(vec![header::CacheDirective::MaxAge(
+        3600,
+    )]));
+    let entity_tag = header::EntityTag::new(false, checksum);
+    response.append_header(header::ETag(entity_tag.clone()));
+    response.append_header(header::ContentLength(out.len()));
+    if let Ok(if_none_match) = <header::IfNoneMatch as header::Header>::parse(&req) {
+        let found_match = match if_none_match {
+            header::IfNoneMatch::Any => true,
+            header::IfNoneMatch::Items(items) => items.iter().any(|etag| etag.weak_eq(&entity_tag)),
+        };
+        if found_match {
+            response.status(actix_web::http::StatusCode::NOT_MODIFIED);
+            return Ok(response.finish());
+        }
+    }
```

Comment:
Would ETags help here? I'd assume it would take longer to actually generate everything than to send it over the network.

Reply:
Etags don't really reduce load on the server, but they do still help reduce network bandwidth/latency. Here's the performance numbers on my computer:

* Generating the feed for the first time after starting labrinth: 0.741 seconds
* Generating a new feed later on, when the database connection has already been cached: 0.007 seconds

Most feeds will obviously be able to use a cached database connection, so should be generated in less than 0.01 seconds, so that's way less time than it takes to send it across the network.